src/NumberTheory/vinberg.jl

```diff
@@ -0,0 +1,292 @@
+end
+
+@doc raw"""
+    _distance_0(Q::ZZMatrix, v0::ZZMatrix, root_lengths::Vector{ZZRingElem}) -> Vector{ZZMatrix}
+
+Return the roots which are orthogonal on v0.
+
+After gathering all vectors v which are orthogonal on v0 with length contained in root_lengths,
+we execute the algorithm again by replacing v0 with the direction vector v1.
+"""
+function _distance_0(Q::ZZMatrix, v0::ZZMatrix, root_lengths::Vector{ZZRingElem}, direction_vector::ZZMatrix)
+  roots = ZZMatrix[]
+  possible_vec = ZZMatrix[]
+  # _short_vectors_gram is more efficient than short_vectors_affine
+  # Thus we have to make some preparations
+  bm = saturate(kernel(Q*transpose(v0); side=:left))
+  QI = bm*Q*transpose(bm)
+  QI = QI[1, 1] > 0 ? QI : -QI
+  for d in root_lengths # gather all vectors which are orthogonal on v0 with length contained in root_lengths 
+    d = abs(d)
+    V = Hecke._short_vectors_gram(Hecke.LatEnumCtx, map_entries(QQ, QI), d, d, ZZRingElem) # QI POSITIVE
+    for (v_, _) in V
+      v = matrix(ZZ, 1, nrows(bm), v_) * bm
+      if !_is_primitive(v)
+        continue
+      end
+      if _crystallographic_condition(Q, v)
+        push!(possible_vec, v)
+      end
+    end
+  end
+  is_empty(possible_vec) && return roots
+  # we check if we can take the given direction vector or generate one if there is no one given
+  if iszero(direction_vector)
+    v1 = _generate_direction_vector(Q, v0, possible_vec)
+  else
+    v1 = _check_direction_vector(Q, v0, possible_vec, direction_vector)
+  end
+  # Now we want to sort the vectors by increase of the value ((v.v1)^2)//v^2
+  possible_vec_sort = Tuple{ZZMatrix,RationalUnion,RationalUnion}[]
+  for v in possible_vec
+    vQ = v * Q
+    n = (vQ*v1)[1, 1]
+    k = (vQ*transpose(v))[1, 1]
+    push!(possible_vec_sort, (v, n, k))
+  end
+  sort!(possible_vec_sort; by=x -> (x[2]^2) // abs(x[3]))
+
+  # We execute the algorithm with replacing v0 by v1
+  for (v, n) in possible_vec_sort
+    if n < 0
+      v = -v
+    end
+    if all((v*Q*transpose(w))[1, 1] >= 0 for w in roots)
+
+      push!(roots, v)
+    end
+  end
+  return roots
+end
+
+@doc raw"""
+    _generate_direction_vector(Q::ZZMatrix, v0::ZZMatrix, possible_vec::Vector{ZZMatrix}) -> ZZMatrix
+
+Since no direction vector v1 was given, we need to find one which satisfies the following conditions:
+- v0.v1 = 0
+- v.v1 != 0 for all possible roots v with v.v0 = 0
+"""
+function _generate_direction_vector(Q::ZZMatrix, v0::ZZMatrix, possible_vec::Vector{ZZMatrix})
+  l = length(v0)
+  v1 = zero_matrix(QQ, l, 1) # initializing v1 that it survives the for loop
+  signal = 1 # initializing a stopping condition
+  while signal in 1:100000000
+    if l < 4
+      v1_ = rand(-30:30, l)
+    elseif l < 10
+      v1_ = rand(-25:25, l)
+    else
+      v1_ = rand(-20:20, l)  # for higher l it is not necessary/efficient to choose a big random range
+    end
+    v1 = matrix(QQ, l, 1, v1_)
+    if (v0*Q*v1)[1, 1] != 0
+      signal += 1
+      continue
+    end
```

Comment:
Why do you impose that v1 and v0 are orthogonal? This should not be necessary. 
Besides it is inefficient to search a vector v1 orthogonal to v0 randomly. This should be done by solving the `ZZ`-linear system of equations instead. 
```suggestion
```